collect_app/src/test/java/org/odk/collect/android/widgets/GeoPointWidgetTest.java

```diff
@@ -1,85 +1,301 @@
+    }
+
+    @Test
+    public void whenPromptHasAnswer_textViewDisplaysAnswer() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
+        String[] parts = answer.split(" ");
+        assertThat(widget.answerDisplay.getText().toString(), equalTo(widget.getContext().getString(
+                R.string.gps_result,
+                GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(widget.getContext(), Double.parseDouble(parts[0]), "lat"),
+                GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(widget.getContext(), Double.parseDouble(parts[1]), "lon"),
+                GeoWidgetUtils.truncateDouble(parts[2]),
+                GeoWidgetUtils.truncateDouble(parts[3])
+        )));
+    }
+
+    @Test
+    public void ifWidgetHasMapsAndIsReadOnly_buttonShowsCorrectText() {
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearance(MAPS, true));
+        assertThat(widget.startGeoButton.getText().toString(), equalTo(widget.getContext().getString(R.string.geopoint_view_read_only)));
+    }
+
+    @Test
+    public void ifWidgetHasMapsAndNullAsAnswer_buttonShowsCorrectText() {
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearanceAndAnswer(MAPS, null));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.get_point)));
+    }
+
+    @Test
+    public void ifWidgetHasMapsAndAnswer_buttonShowsCorrectText() {
+        when(mapConfigurator.isAvailable(any())).thenReturn(true);
+        GeoPointWidget widget = createWidget(promptWithAppearanceAndAnswer(MAPS, new StringData(answer)));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.view_change_location)));
+    }
+
+    @Test
+    public void whenWidgetHasAnswer_buttonShowsCorrectText() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.change_location)));
+    }
+
+    @Test
+    public void whenWidgetHasNullAsAnswer_buttonShowsCorrectText() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(widget.startGeoButton.getText(), equalTo(widget.getContext().getString(R.string.get_point)));
+    }
+
+    @Test
+    public void whenPermissionIsNotGranted_buttonShouldNotLaunchAnyIntent() {
+        GeoPointWidget widget = createWidget(promptWithAnswer(null));
+        stubLocationPermissions(widget, false);
+        widget.startGeoButton.performClick();
+        Intent startedIntent = shadowOf(widgetTestActivity()).getNextStartedActivity();
+
+        assertNull(startedIntent);
+    }
+
+    @Test
+    public void whenPermissionIsGranted_buttonClickShouldSetTheLocationSavedOnTheMap() {
```

Comment:
A more accurate name might be `buttonClickLaunchesIntentAndWaitsForLocationData`.